File: src/test/java/no/nav/fo/veilarbdialog/brukernotifikasjon/kvittering/EksternVarslingKvitteringConsumerTest.java
package no.nav.fo.veilarbdialog.brukernotifikasjon.kvittering;

import no.nav.doknotifikasjon.schemas.DoknotifikasjonStatus;
import no.nav.fo.veilarbdialog.brukernotifikasjon.BrukernotifikasjonRepository;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.UUID;

import static org.mockito.Mockito.*;

/**
 * Tester noen caser som er vrien å teste med integrasjonstester (typisk async feilhåndtering)
 */
@ExtendWith(MockitoExtension.class)
class EksternVarslingKvitteringConsumerTest {

    private static final String APP_NAME = "veilarbdialog";

    EksternVarslingKvitteringConsumer eksternVarslingKvitteringConsumer;

    @Mock
    private KvitteringDAO kvitteringDAO;
    @Mock
    private BrukernotifikasjonRepository brukernotifikasjonRepository;

    @Mock
    private KvitteringMetrikk kvitteringMetrikk;

    @BeforeEach
    void setup() {
        eksternVarslingKvitteringConsumer = new EksternVarslingKvitteringConsumer(kvitteringDAO, brukernotifikasjonRepository, kvitteringMetrikk, APP_NAME);
    }

    @Test
    void skalIgnorereMeldingerMedAnnenBestillerid() {
        String feilApp = "annen-app";
        String bestillingsId = UUID.randomUUID().toString();
        DoknotifikasjonStatus doknotifikasjonStatus = DoknotifikasjonStatus
                .newBuilder()
                .setStatus("INFO")
                .setBestillingsId(bestillingsId)
                .setBestillerId(feilApp)
                .setMelding("her er en melding")
                .setDistribusjonId(1L)
                .build();

        ConsumerRecord consumerRecord = createConsumerRecord(doknotifikasjonStatus);
        eksternVarslingKvitteringConsumer.consume(consumerRecord);
        verifyNoInteractions(
                kvitteringDAO, brukernotifikasjonRepository, kvitteringMetrikk);
    }

    @Test
    void skalFeileVedUkjentStatus() {
        String bestillingsId = UUID.randomUUID().toString();
        DoknotifikasjonStatus doknotifikasjonStatus = createDoknotifikasjonStatus(bestillingsId, "RUBBISH");

        when(brukernotifikasjonRepository.finnesBrukernotifikasjon(bestillingsId)).thenReturn(true);

        ConsumerRecord consumerRecord = createConsumerRecord(doknotifikasjonStatus);
        Assertions.assertThrows(IllegalArgumentException.class,
                () -> eksternVarslingKvitteringConsumer.consume(consumerRecord));

        verify(kvitteringDAO).lagreKvittering(bestillingsId, doknotifikasjonStatus);

        verify(brukernotifikasjonRepository).finnesBrukernotifikasjon(bestillingsId);
        verifyNoMoreInteractions(brukernotifikasjonRepository);
    }

    @Test
    void skalFeileVedUkjentBestillingsId() {
        String bestillingsId = UUID.randomUUID().toString();
        DoknotifikasjonStatus doknotifikasjonStatus = createDoknotifikasjonStatus(bestillingsId, "OVERSENDT");

        when(brukernotifikasjonRepository.finnesBrukernotifikasjon(bestillingsId)).thenReturn(false);


        ConsumerRecord consumerRecord = createConsumerRecord(doknotifikasjonStatus);
        Assertions.assertThrows(IllegalArgumentException.class,
                () -> eksternVarslingKvitteringConsumer.consume(consumerRecord));
        verify(brukernotifikasjonRepository).finnesBrukernotifikasjon(bestillingsId);
        verifyNoMoreInteractions(brukernotifikasjonRepository);
    }

    private ConsumerRecord createConsumerRecord(DoknotifikasjonStatus doknotifikasjonStatus) {
        return new ConsumerRecord("kvitteringsTopic", 0, 1, doknotifikasjonStatus.getBestillingsId(), doknotifikasjonStatus);
    }

    private DoknotifikasjonStatus createDoknotifikasjonStatus(String bestillingsId, String status) {
        return DoknotifikasjonStatus
                .newBuilder()
                .setStatus(status)
                .setBestillingsId(bestillingsId)
                .setBestillerId(APP_NAME)
                .setMelding("her er en melding")
                .setDistribusjonId(1L)
                .build();
    }
}

File: src/test/java/no/nav/fo/veilarbdialog/brukernotifikasjon/kvittering/EksternVarslingKvitteringConsumerTest.kt
package no.nav.fo.veilarbdialog.brukernotifikasjon.kvittering

import java.util.UUID
import no.nav.common.json.JsonUtils
import no.nav.fo.veilarbdialog.brukernotifikasjon.kvittering.VarselHendelseUtil.eksternVarselHendelseSendt
import no.nav.fo.veilarbdialog.minsidevarsler.MinsideVarselHendelseConsumer
import no.nav.fo.veilarbdialog.minsidevarsler.MinsideVarselService
import no.nav.fo.veilarbdialog.minsidevarsler.dto.MinSideVarselId
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension
import tools.jackson.databind.ObjectMapper
import tools.jackson.databind.node.ObjectNode
import tools.jackson.databind.node.StringNode

/**
 * Tester noen caser som er vrien å teste med integrasjonstester (typisk async feilhåndtering)
 */
@ExtendWith(MockitoExtension::class)
open class EksternVarslingKvitteringConsumerTest(
    @Mock
    private val minsideVarselService: MinsideVarselService,
    @Mock
    private val kvitteringMetrikk: KvitteringMetrikk,
) {

    var eksternVarslingKvitteringConsumer: MinsideVarselHendelseConsumer? = null

    companion object {
        val minsideVarselHendelseTopic = "topic"
        private const val APP_NAME = "veilarbdialog"
    }

    @BeforeEach
    fun setup() {
        eksternVarslingKvitteringConsumer = MinsideVarselHendelseConsumer(APP_NAME, minsideVarselService, kvitteringMetrikk)
    }

    @Test
    fun skalIgnorereMeldingerMedAnnenBestillerid() {
        val feilApp = "annen-app"
        val varselHendelse = eksternVarselHendelseSendt(MinSideVarselId(UUID.randomUUID()), feilApp)

        val consumerRecord = createConsumerRecord(varselHendelse)
        eksternVarslingKvitteringConsumer!!.consume(consumerRecord)
        Mockito.verifyNoInteractions(minsideVarselService, kvitteringMetrikk)
    }

        @Test
        fun `skal ikke interagere med minsideVarselService ved ukjent status`() {
            var varselId = MinSideVarselId(UUID.randomUUID())
            val hendelseMedUkjentStatus = eksternVarselHendelseSendt(varselId, APP_NAME)
                .let(JsonUtils::toJson)
                .let { ObjectMapper().readTree(it) as ObjectNode }
                .set("status", StringNode("lol"))
                .toString()
            val consumerRecord = ConsumerRecord<String, String>(
                minsideVarselHendelseTopic, 0, 1,
                varselId.value.toString(),
                hendelseMedUkjentStatus
            )
            // When
            Assertions.assertThrows(IllegalArgumentException::class.java) {
                eksternVarslingKvitteringConsumer!!.consume(
                    consumerRecord
                )
            }
            // Then
            Mockito.verifyNoMoreInteractions(minsideVarselService);
        }

    private fun createConsumerRecord(varselHendelse: TestVarselHendelseDTO): ConsumerRecord<String, String> {
        return ConsumerRecord<String, String>(
            minsideVarselHendelseTopic,
            0,
            1,
            varselHendelse.varselId.toString(),
            JsonUtils.toJson(varselHendelse)
        )
    }
}